flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/CodeGenUtils.scala

```diff
@@ -0,0 +1,176 @@
+  }
+
+  def primitiveDefaultValue(tpe: TypeInformation[_]): String = tpe match {
+    case INT_TYPE_INFO => "-1"
+    case LONG_TYPE_INFO => "-1"
+    case SHORT_TYPE_INFO => "-1"
+    case BYTE_TYPE_INFO => "-1"
+    case FLOAT_TYPE_INFO => "-1.0f"
+    case DOUBLE_TYPE_INFO => "-1.0d"
+    case BOOLEAN_TYPE_INFO => "false"
+    case STRING_TYPE_INFO => "\"<empty>\""
+    case CHAR_TYPE_INFO => "'\\0'"
+    case _ => "null"
+  }
+
+  def requireNumeric(genExpr: GeneratedExpression) = genExpr.resultType match {
+    case nti: NumericTypeInfo[_] => // ok
+    case _ => throw new CodeGenException("Numeric expression type expected.")
+  }
+
+  def requireString(genExpr: GeneratedExpression) = genExpr.resultType match {
+    case STRING_TYPE_INFO => // ok
+    case _ => throw new CodeGenException("String expression type expected.")
+  }
+
+  def requireBoolean(genExpr: GeneratedExpression) = genExpr.resultType match {
+    case BOOLEAN_TYPE_INFO => // ok
+    case _ => throw new CodeGenException("Boolean expression type expected.")
+  }
+
+  def isReference(genExpr: GeneratedExpression): Boolean = isReference(genExpr.resultType)
+
+  def isReference(typeInfo: TypeInformation[_]): Boolean = typeInfo match {
+    case INT_TYPE_INFO
+         | LONG_TYPE_INFO
+         | SHORT_TYPE_INFO
+         | BYTE_TYPE_INFO
+         | FLOAT_TYPE_INFO
+         | DOUBLE_TYPE_INFO
+         | BOOLEAN_TYPE_INFO
+         | CHAR_TYPE_INFO => false
+    case _ => true
+  }
+
+  def isNumeric(genExpr: GeneratedExpression): Boolean = isNumeric(genExpr.resultType)
+
+  def isNumeric(typeInfo: TypeInformation[_]): Boolean = typeInfo match {
+    case nti: NumericTypeInfo[_] => true
+    case _ => false
+  }
+
+  def isString(genExpr: GeneratedExpression): Boolean = isString(genExpr.resultType)
+
+  def isString(typeInfo: TypeInformation[_]): Boolean = typeInfo match {
+    case STRING_TYPE_INFO => true
+    case _ => false
+  }
+
+  def isBoolean(genExpr: GeneratedExpression): Boolean = isBoolean(genExpr.resultType)
+
+  def isBoolean(typeInfo: TypeInformation[_]): Boolean = typeInfo match {
+    case BOOLEAN_TYPE_INFO => true
+    case _ => false
+  }
+
+  // ----------------------------------------------------------------------------------------------
+
+  sealed abstract class FieldAccessor
+
+  case class ObjectFieldAccessor(fieldName: String) extends FieldAccessor
+
+  case class ObjectMethodAccessor(methodName: String) extends FieldAccessor
+
+  case class ProductAccessor(i: Int) extends FieldAccessor
+
+  def fieldAccessorFor(compType: CompositeType[_], index: Int): FieldAccessor = {
+    compType match {
+      case ri: RowTypeInfo =>
+        ProductAccessor(index)
+
+      case cc: CaseClassTypeInfo[_] =>
+        ObjectMethodAccessor(cc.getFieldNames()(index))
+
+      case javaTup: TupleTypeInfo[_] =>
+        ObjectFieldAccessor("f" + index)
+
+      case pj: PojoTypeInfo[_] =>
+        ObjectFieldAccessor(pj.getFieldNames()(index))
```

Comment:
PojoFields may be private. 
You need to use reflection and previously make the field accessible.